File: rust_version/src/core/memory_pool.rs
use std::alloc::{alloc, dealloc, Layout};
use std::ptr::NonNull;
use std::sync::Mutex;

pub struct MemoryPool {
    block_size: usize,
    free_list: Mutex<Option<NonNull<Block>>>,
    chunks: Mutex<Vec<NonNull<u8>>>,
}

struct Block {
    next: Option<NonNull<Block>>,
}

impl MemoryPool {
    pub fn new(block_size: usize, initial_blocks: usize) -> Self {
        let pool = MemoryPool {
            block_size: block_size.max(std::mem::size_of::<Block>()),
            free_list: Mutex::new(None),
            chunks: Mutex::new(Vec::new()),
        };
        pool.allocate_chunk(initial_blocks);
        pool
    }

    pub fn allocate(&self) -> NonNull<u8> {
        let mut free_list = self.free_list.lock().unwrap();
        if free_list.is_none() {
            drop(free_list);  // 释放锁，因为 allocate_chunk 会再次获取锁
            self.allocate_chunk(100);
            free_list = self.free_list.lock().unwrap();
        }
        let block = free_list.take().unwrap();
        *free_list = unsafe { block.as_ref().next };
        block.cast()
    }

    pub fn deallocate(&self, ptr: NonNull<u8>) {
        let block = ptr.cast::<Block>();
        let mut free_list = self.free_list.lock().unwrap();
        unsafe {
            (*block.as_ptr()).next = *free_list;
        }
        *free_list = Some(block);
    }

    fn allocate_chunk(&self, blocks: usize) {
        let layout = Layout::array::<u8>(self.block_size * blocks).unwrap();
        let chunk = unsafe { NonNull::new(alloc(layout)).unwrap() };
        
        let mut free_list = self.free_list.lock().unwrap();
        for i in 0..blocks {
            let block = unsafe { chunk.as_ptr().add(i * self.block_size) as *mut Block };
            let block = NonNull::new(block).unwrap();
            unsafe {
                (*block.as_ptr()).next = *free_list;
            }
            *free_list = Some(block);
        }

        self.chunks.lock().unwrap().push(chunk);
    }
}

impl Drop for MemoryPool {
    fn drop(&mut self) {
        let chunks = self.chunks.get_mut().unwrap();
        for &chunk in chunks.iter() {
            unsafe {
                dealloc(chunk.as_ptr(), Layout::array::<u8>(self.block_size * 100).unwrap());
            }
        }
    }
}

File: cpp_version/src/core/memory_pool.cpp
#include "core/memory_pool.h"
#include <cstdlib>
#include <algorithm>

MemoryPool::MemoryPool(size_t block_size, size_t initial_blocks)
    : block_size_(std::max(block_size, sizeof(Block))), 
      blocks_per_chunk_(initial_blocks),free_list_(nullptr) 
{
    allocate_chunk();
}

MemoryPool::~MemoryPool() {
    for (auto chunk : chunks_) {
        std::free(chunk);
    }
}

void* MemoryPool::allocate() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (free_list_ == nullptr) {
        allocate_chunk();
    }

    Block* block = free_list_;
    free_list_ = block->next;
    return block;
}

void MemoryPool::deallocate(void* ptr) {
    std::lock_guard<std::mutex> lock(mutex_);
    Block* block = static_cast<Block*>(ptr);
    block->next = free_list_;
    free_list_ = block;
}

void MemoryPool::allocate_chunk(){
    size_t chunk_size = block_size_ * blocks_per_chunk_;
    char* chunk = static_cast<char*>(std::malloc(chunk_size));
    chunks_.push_back(chunk);

    for(size_t i = 0; i< chunk_size; i+=block_size_){
        Block* block = reinterpret_cast<Block*>(chunk + i);
        block->next = free_list_;
        free_list_ = block;
    }
}